source/channel_bootstrap.c

```diff
@@ -36,6 +37,55 @@
 #define MAX_HOST_RESOLVER_ENTRIES 64
 #define DEFAULT_DNS_TTL 30
 
+struct tl_shutdown_task_data {
+    struct aws_condition_variable *condition_variable;
+    struct aws_mutex *mutex;
+    bool invoked;
+};
+
+static bool s_tl_cleanup_predicate(void *arg) {
+    struct tl_shutdown_task_data *shutdown_task_data = arg;
+    return shutdown_task_data->invoked;
+}
+
+static void s_handle_tl_cleanup_task(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    (void)status;
+    struct tl_shutdown_task_data *shutdown_task_data = arg;
+
+    aws_mutex_lock(shutdown_task_data->mutex);
+    aws_tls_clean_up_thread_local_state();
+    shutdown_task_data->invoked = true;
+    aws_condition_variable_notify_one(shutdown_task_data->condition_variable);
+    aws_mutex_unlock(shutdown_task_data->mutex);
+}
+
+static void s_ensure_tl_state_is_cleaned_up(struct aws_event_loop_group *el_group) {
+    struct aws_mutex mutex = AWS_MUTEX_INIT;
+    struct aws_condition_variable condition_variable = AWS_CONDITION_VARIABLE_INIT;
+
+    aws_mutex_lock(&mutex);
+    size_t len = aws_event_loop_group_get_loop_count(el_group);
+    for (size_t i = 0; i < len; ++i) {
+        struct aws_event_loop *el = aws_event_loop_group_get_loop_at(el_group, i);
+
+        struct tl_shutdown_task_data tl_shutdown = {
+            .mutex = &mutex,
+            .condition_variable = &condition_variable,
+            .invoked = false,
+        };
+
+        struct aws_task task = {
+            .fn = s_handle_tl_cleanup_task,
+            .arg = &tl_shutdown,
+        };
+
+        aws_event_loop_schedule_task_now(el, &task);
+        aws_condition_variable_wait_pred(&condition_variable, &mutex, s_tl_cleanup_predicate, &tl_shutdown);
+    }
+    aws_mutex_unlock(&mutex);
```

Comment:
I don't think this is accurate.
1st, by definition with condition variable wait will unlock the mutex when it waits. Since it's in a different thread, this guarantees the task lock isn't acquired until we try to wait.
2nd, two scenarios, if predicate prevents the wait by returning true, the mutex was never unlocked, otherwise, the condition var wait will reaquire the lock before returning, hence the needed unlock. 

I did move the unlock in the task above the notify though.